src/include/column_inline.h

```diff
@@ -19,9 +19,10 @@ __col_insert_search_gt(WT_INSERT_HEAD *ins_head, uint64_t recno)
     /*
      * Compiler may replace the following usage of the variable with another read.
      *
-     * Place a read barrier to avoid this issue.
+     * Place a compiler barrier to avoid this issue.
      */
-    WT_ORDERED_READ_WEAK_MEMORDER(ins, WT_SKIP_LAST(ins_head));
+    ins = WT_SKIP_LAST(ins_head);
+    WT_C11_BARRIER();
```

Comment:
If we are worried about this, we should make `ins->tail` atomic instead of using a barrier.

Reply:
I don't really follow. line 32 is reading through a pointer we already got. Can it be reordered before we actually read the pointer `ins = WT_SKIP_LAST(ins_head)`?

Here we only intend to not read WT_SKIP_LAST(ins_head) to ins multiple times. I don't think we need to worry about the order as there is a single thread happen-before relationship?

Reply:
> line 32 is reading through a pointer we already got. Can it be reordered before we actually read the pointer ins = WT_SKIP_LAST(ins_head).

Yes. Well, more precisely, the read through the pointer isn't guaranteed to happen after the write to that location unless you have a proper chain of sequencing like the following:

1. `*producer_private = value;`
2. `public_ptr.store(producer_private, release);`
3. `consumer_private = public_ptr.load(acquire);` where that reads the value written in step 2 (or something in it release sequence)
4. `value = *consumer_private;`

What we really care about is that step 1 _happens before_ step 4. In the C memory model, that is ensured by having step 3 do a load-acquire that reads the value written by step 2. See http://eel.is/c++draft/atomics.order#2. If you want a short hard way to remember and reason about this rule, remember that acquire and release operations don't do anything on their own. They only "activate" in pairs, and only for the pair of threads involved in the pairing (or others that transitively establish an ordering with them through additional operations). 

Note that using `seq_cst` doesn't change this, it just adds a rule that in addition to the acquire-release ordering, there is a single total ordering between all `seq_cst` operations, which means that `seq_cst` is only stronger than acquire/release when paired with other `seq_cst` opertaions.

Reply:
PS - what I think you are describing is something that hardware often guarantees (with the notable exception of the DEC Alpha) under the name "address dependency ordering".  It basically says that any reads that contribute to the value of an address must happen before a read of that address. Arm also provides this guarantee at the assembly level, however it has weird limitations around transitivity relative to using acquire loads as can be seen [here](https://wg21.link/p0735r1).

There was an attempt to provide access to this guarantee in the C memory model as `memory_order_consume`, however compiler implementors have found it basically impossible to preserve the guarantees from source code all the way to generated assembly so they all just implement it with the same codegen as acquire, which is why I say it is usually best to pretend that consume doesn't exist. There have been several attempts at fixing consume for c++ (see [P0750](https://wg21.link/p0750) along with the links in its background section), but AFAICT there hasn't been much movement on that front recently). The LKMM (linux kernel memory model) tries to allow developers to take advantage of this hardware guarantee, but you can see all of the caveats that they document [here](https://github.com/torvalds/linux/blob/master/Documentation/RCU/rcu_dereference.rst). I do not recommend we attempt to duplicate this, unless we find some *really* hot code path where even `LDAPR` is too expensive.

Reply:
Then I think in addition to an acquire read here, we also need a release store when we write to the head.

Reply:
> Then I think in addition to an acquire read here, we also need a release store when we write to the head.

Exactly! Do we not have that right now?

Reply:
We have that for the insert list head as it is changed by compare and swap but we miss that in the tail. I fixed that in the pr.

Reply:
> We have that for the insert list head as it is changed by compare and swap but we miss that in the tail. I fixed that in the pr.

Are you referring to [this](https://github.com/wiredtiger/wiredtiger/pull/9010/commits/1ae6a0bc084fa9e65ae8df99605e6b5ea862498a#diff-302b65ca76325e7e93eb73d95d1993931e1b90d3a660d8a5fd4417c7d005ae3eR122-R127) change? If so, I think that looks like a bug even under the "old" WT memory model, and should be fixed in master, separately from this PR.

PS- finding these sorts of things is why I think that thinking in terms of finding the paired acquires and releases helps to avoid introducing memory ordering bugs, rather than just thinking about the barriers needed for each side in isolation.

Reply:
That's correct. However, I'm not sure that is actually a problem in the original wt code as anyway we never use an acquire read to read the head or tail. I think the assumption is that it is ok to read an inconsistent value from head or tail as they are the places we start walking the list and missing a few new items is ok. I changed this in the pr to make the usage of the C11 model complete.

Reply:
We _may_ also be saved on master by the fact that the immediately prior write was a CAS (implying a full barrier on master), so this store couldn't be reordered any earlier, even without a separate store barrier in front. But I find these sorts of "accidentally correct" cases a bit hard to reason about.